### PR TITLE
First draft of http package

### PR DESCRIPTION
**Pull request by markstory**
I've had to move an exception around and add a new interface to allow separation between the controller and http interfaces. While there is one reference to `Router` in `ActionDispatcher` the `class_exists()` check should make it an optional dependency.

### CODE REVIEW

File: src/Http/composer.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            "name": "CakePHP Community",
+            "homepage": "https://github.com/cakephp/form/graphs/contributors"
+        }
+    ],
+    "support": {
+        "issues": "https://github.com/cakephp/cakephp/issues",
+        "forum": "https://stackoverflow.com/tags/cakephp",
+        "irc": "irc://irc.freenode.org/cakephp",
+        "source": "https://github.com/cakephp/form"
```
Comment (saeideng) [inline src/Http/composer.json]:
```suggestion
        "source": "https://github.com/cakephp/http"
```

File: src/Http/composer.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        "http",
+        "psr7",
+        "psr15"
+    ],
+    "homepage": "https://cakephp.org",
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "CakePHP Community",
+            "homepage": "https://github.com/cakephp/form/graphs/contributors"
```
Comment (saeideng) [inline src/Http/composer.json]:
```suggestion
            "homepage": "https://github.com/cakephp/http/graphs/contributors"
```

File: src/Http/ControllerInterface.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+     * @throws \UnexpectedValueException If return value of action method is not null or ResponseInterface instance.
+     */
+    public function invokeAction(): ?ResponseInterface;
+
+    /**
+     * Returns true if an action should be rendered automatically.
+     *
+     * @return bool
+     */
+    public function isAutoRenderEnabled(): bool;
```
Comment (ADmad) [inline src/Http/ControllerInterface.php]: This method feels quite out of place on the interface.

Comment (markstory) [reply src/Http/ControllerInterface.php]: I agree but the ActionDispatcher does call this method so it was necessary to get typing to work.

File: src/Http/Exception/MissingControllerException.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Constructor.
+     *
+     * @param string|array $message Either the string of the error message, or an array of attributes
+     *   that are made available in the view, and sprintf()'d into Exception::$_messageTemplate
+     * @param int|null $code The code of the error, is also the HTTP status code for the error. Defaults to 404.
+     * @param \Throwable|null $previous the previous exception.
+     */
+    public function __construct($message, ?int $code = 404, ?Throwable $previous = null)
```
Comment (garas) [inline src/Http/Exception/MissingControllerException.php]: I've looked why there is `404` here, but `500` in `ExceptionsTest::exceptionProvider()`.

It should be no `__construct()` here, but it should be `protected $_defaultCode = 404;` instead. With second argument as `null`, it fallbacks to `500`.

Core exceptions has mixed between `__construct()` and `protected $_defaultCode`, they should be probably normalized in next PR.

File: src/Http/ControllerInterface.php (hunk 3 above)
Comment (ADmad) [reply src/Http/ControllerInterface.php]: Couldn't the called checking using `isAutoRenderEnabled()` and the rendering be moved inside `Controller::invokeAction()` itself?

Comment (markstory) [reply src/Http/ControllerInterface.php]: Yes, that could work. We could move the isAutoRenderEnabled() and render calls there.

Comment (markstory) [reply src/Http/ControllerInterface.php]: Done :+1:

**Comment by markstory** [comment]
@ADmad Anything else you'd like to see here?

File: src/Http/ControllerInterface.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * Dispatches the controller action. Checks that the action
+     * exists and isn't private.
+     *
+     * @return \Psr\Http\Message\ResponseInterface The resulting response.
+     * @throws \Cake\Controller\Exception\MissingActionException If controller action is not found.
+     * @throws \UnexpectedValueException If return value of action method is not null or ResponseInterface instance.
+     */
+    public function invokeAction(): ?ResponseInterface;
```
[the remark this answers is not in the document]
Comment (markstory) [reply src/Http/ControllerInterface.php]: If we normalize on ResponseInterface, linters will complain about the methods we've added. We should probably use Response.

[the remark this answers is not in the document]
Comment (ADmad) [reply src/Http/ControllerInterface.php]: Okay, let's keep the return type as they are for now.

[the remark this answers is not in the document]
Comment (markstory) [reply src/Http/ControllerInterface.php]: Done.

**Pull request merged**